crypto/params.c

```diff
@@ -1562,79 +1566,60 @@ int ossl_param_get1_octet_string(const OSSL_PARAM *params, const char *name,
     return 1;
 }
 
-static int setbuf_fromparams(const OSSL_PARAM *p, const char *name,
-                             unsigned char *out, size_t *outlen)
-{
-    int ret = 0;
-    WPACKET pkt;
-
-    if (out == NULL) {
-        if (!WPACKET_init_null(&pkt, 0))
-            return 0;
-    } else {
-        if (!WPACKET_init_static_len(&pkt, out, *outlen, 0))
-            return 0;
-    }
-
-    for (; p != NULL; p = OSSL_PARAM_locate_const(p + 1, name)) {
-        if (p->data_type != OSSL_PARAM_OCTET_STRING)
-            goto err;
-        if (p->data != NULL
-                && p->data_size != 0
-                && !WPACKET_memcpy(&pkt, p->data, p->data_size))
-            goto err;
-    }
-    if (!WPACKET_get_total_written(&pkt, outlen)
-            || !WPACKET_finish(&pkt))
-        goto err;
-    ret = 1;
-err:
-    WPACKET_cleanup(&pkt);
-    return ret;
-}
-
 int ossl_param_get1_concat_octet_string(const OSSL_PARAM *params, const char *name,
                                         unsigned char **out,
                                         size_t *out_len, size_t maxsize)
 {
     const OSSL_PARAM *p = OSSL_PARAM_locate_const(params, name);
     unsigned char *res;
+    size_t res_len = *out_len;
+    size_t cur_len = *out_len;
+    int err = 0;
+    const void *val = NULL;
     size_t sz = 0;
 
     if (p == NULL)
         return -1;
 
-    /* Calculate the total size */
-    if (!setbuf_fromparams(p, name, NULL, &sz))
```

Comment:
Not quite sure why this needed to be rewritten.
The basic gist of the problem seems to be that the existing code just seems to concat all the newly added octet strings with the same name together.
What it doesnt do is join the existing data with the newly added octet string data..
I do think WPACKET is a much safer way of joining things together. Then if you use the same logic to calculate the size there is less room for error.
Using OSSL_PARAM_get_octet_string_ptr() is better

Reply:
I'm not sure `WPACKET` is the way to go.

Reply:
The use of `WPACKET` means that it's allocating more than once: when determining the size, it could allocate multiple times. On the second pass, data is allocated again. So, (re)allocations could occur multiple times, which is not efficient.

Reply:
I dont think this is true.
On the first pass it passed NULL as the buffer (using WPACKET_init_null), so it shouldnt allocate a buffer at all since it doesnt need one (i.e it just increments when doing WPACKET_memcpy()). On the second pass it passes the buffer of the total size that is needs which should allocate the required space, so it only does the one alloc.